Compute particular commit order for deletions

jusephe opened the pull request:
Fixes: https://github.com/doctrine/orm/issues/9376

jusephe commented:
I just pushed new commit to fix "Static Analysis with Psalm".

SenseException commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -444,8 +444,14 @@ public function commit($entity = null)
             if ($this->entityDeletions) {
-                for ($count = count($commitOrder), $i = $count - 1; $i >= 0 && $this->entityDeletions; --$i) {
-                    $this->executeDeletions($commitOrder[$i]);
+                // There are some cases in which the original $commitOrder is not correct for deletions,
+                // so we need to call a particular function to compute the right order.
+                // For example see tests/Doctrine/Tests/ORM/Functional/Ticket/GH9376Test.php.
+                // Without computing the special order for deletions, it tries to delete the entity set as a foreign key in another entity.
+                // getCommitOrderForDeletions() is similar to getCommitOrder(), but it operates only with entities set to be deleted.
```
This should be part of the description of a commit message. Your method name seem to describe well what is happening.

jusephe replied:
fixed

SenseException commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -1330,6 +1336,54 @@ private function getCommitOrder(): array
+    {
+        $calc = $this->getCommitOrderCalculator();
+
+        // See if there are any new classes in the changeset, that are not in the
+        // commit order graph yet (don't have a node).
+        // We have to inspect changeSet to be able to correctly build dependencies.
+        // It is not possible to use IdentityMap here because post inserted ids
+        // are not yet available.
```
```suggestion
```

jusephe replied:
fixed

SenseException commented:
It looks like your test does not cover your fix, because when I undo your changes in UnitOfWork, the test doesn't fail. ORM seems to be able to handle the test case.

jusephe commented:
> 

@SenseException Have you tried with a DB platform supporting foreign key constraints. (i.e., not SQLite)? I tried it again and it still fails...

jusephe commented:
@SenseException Can I help you somehow with testing the issue? I rebased my branch onto current version and the test still fails if I undo my changes. The test has to be run with a DB platform supporting foreign key constraints. (i.e., NOT SQLite).

SenseException commented:
Sorry for the late answer and for forgetting about this PR. The commit order calculation isn't an easy topic and I'll have to take more time to review it. I also want to prevent possible impacts on performance.

SenseException commented:
Please change the base branch to 2.12.x and rebase.

jusephe commented:
@SenseException I rebased my branch onto 2.12.x. The test still fails if I undo my changes. (The test has to be run with a DB platform supporting foreign key constraints (i.e., NOT SQLite).

SenseException commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -1331,6 +1332,49 @@ private function getCommitOrder(): array
+
+        $newNodes = [];
+
+        foreach ($this->entityDeletions as $entity) {
```
With `$this->entityDeletions` in here, it wouldn't be needed in `$this->getCommitOrder()` anymore. Otherwise the commit order would calculate a second time using the deletions.

SenseException commented on tests/Doctrine/Tests/ORM/Functional/Ticket/GH9376Test.php:
```diff
@@ -0,0 +1,204 @@
+        parent::tearDown();
+    }
+
+    public function testIssueRemove(): void
```
```suggestion
    public function testRemoveCircularRelatedEntities(): void
```

SenseException reviewed (commented):
@doctrine/doctrinecore

jusephe commented:
@SenseException Can I help you somehow with this PR? I rebased my branch onto 2.13.x and the test still fails if I undo my changes.

jusephe committed:
Compute particular commit order for deletions

This fixes a bug which arises in some cases when trying to delete two entities which have a circular reference through other entities. UnifOfWork computed the order which was not correct for deletions and it failed with ForeignKeyConstraintViolationException. To fix that, this adds a new particular function getCommitOrderForDeletions() to compute the right order. This function is similar to getCommitOrder(), but it operates only with entities set to be deleted. This also adds test for this issue.

SenseException commented:
Thank you for your offer and your patience. There's currently no help required that I know of. We'll come back in case this PR needs changes or maybe new test cases to prevent breaking changes.

mpdude commented:
This will be fixed by #10547, make sure to track that one.

mpdude commented:
Also for me, the test case passes on the current 2.16.x branch without further modifications to the UoW. Yes, I am using a DBMS with foreign keys.

Could you please re-check that and/or provide only the failing test in a dedicated PR against 2.16 (and notify me)?

Thanks!

/cc @greg0ire

mpdude commented:
@jusephe are you still interested in working on this?

jusephe commented:
@mpdude Hi, sorry for late answer. The problem is fixed either by https://github.com/doctrine/orm/pull/10547 or https://github.com/doctrine/orm/pull/10913. Thank you very much! I am closing this.
